### ISSUE TITLE
Next Track Not Working on iOS 9.2. Was working with iOS 9.1

### ISSUE THREAD

**Issue by jojomama1995** [mikebrady/shairport-sync#177]
I having issues skipping to the next track on iOS 9.2. Can someone help confirm that they are experiencing this issue or not or if I'm just crazy? Thank you

**Comment by mikebrady** [comment]
Thanks for your report. Have you enabled the fetching of metadata?

**Comment by jojomama1995** [comment]
I'm not familiar with fetching of metadata, so if it's something that I need to explicitly enable, then I'm not using it.

The issue seem to be related to iOS 9.2 which was released last week.

**Comment by mikebrady** [comment]
No need to worry! Metadata is extra stuff like pictures and titles. It's not relevant to a simple player, so you should leave it switched off. The thing is that if it was enabled, sometimes the amount of metadata can be very large and slightly delay commands, but not audio. Even worse, metadata can sometimes stall if it's coming from iTunes. In your case, it's not enabled so this shouldn't be an issue, and anyway your source isn't iTunes, it's iOS.

So, your issue remains a mystery – I haven't come across it. Maybe somebody else has...

**Comment by mikebrady** [comment]
Hi there. Was there anything new on this?

**Comment by KuneStudio** [comment]
I'm following the issue. I also have that problem, and need the metadata to be displayed on a screen. Any news ? Thanks ! :)

**Comment by mikebrady** [comment]
Hi @KuneStudio. Could you describe your problem more precisely please? Do you already have a metadata feed to a screen?

**Comment by KuneStudio** [comment]
Yes I can.
Using shairport on a vanilla install of raspian and with RuneAudio on another RPI.
On both, when using the next button on Music app (IOS 9.2), it causes a deconnection of the airplay.
When using Spotify, the next button is working. 
Yes, I have a feed on a screen, but it looks like it's not related. the second rpi with the vanilla install and shairport are not set up to use the screen.

**Comment by mikebrady** [comment]
Thanks. When selecting the next track on the Music app, is the track locally stored in the iOS device, or is it in the cloud, e.g. in iTunes Match or in the Apple Music service? In other words, I wonder if it makes a difference if the next track is local or remote... I'll see if I can replicate it.

**Comment by KuneStudio** [comment]
Oh I did not check that but it seems that the problem is coming when the first music is on the cloud and the other one is local (and the other way too). It's not happening everytime, but it does happen a lot.

**Comment by mikebrady** [comment]
Thanks.

**Comment by joerg-krause** [comment]
I am experiencing the same issue. It happens a lot if I press the next button two or three times, to skip more than one track.

**Comment by joerg-krause** [comment]
This issue also happens if I skip one track. Here is a log (WM8524 parts are coming from the codec when it starts playing):

```
[  800.408116] WM8524: Sampling Rate=44100Hz
[  800.412195] WM8524: MCLK=16.934MHz
Dropping flushed packet in player_put_packet, seqno 59576, timestamp 2963654171, flushing to timestamp: 2963658009.
Dropping flushed packet in player_put_packet, seqno 59577, timestamp 2963654523, flushing to timestamp: 2963658009.
RTP: Packets out of sequence: expected: 59578, got 59579.
Dropping flushed packet in player_put_packet, seqno 59579, timestamp 2963655227, flushing to timestamp: 2963658009.
Dropping flushed packet in player_put_packet, seqno 59580, timestamp 2963655579, flushing to timestamp: 2963658009.
Dropping flushed packet in player_put_packet, seqno 59581, timestamp 2963655931, flushing to timestamp: 2963658009.
Dropping flushed packet in player_put_packet, seqno 59582, timestamp 2963656283, flushing to timestamp: 2963658009.
Dropping flushed packet in player_put_packet, seqno 59583, timestamp 2963656635, flushing to timestamp: 2963658009.
Dropping flushed packet in player_put_packet, seqno 59584, timestamp 2963656987, flushing to timestamp: 2963658009.
Dropping flushed packet in player_put_packet, seqno 59585, timestamp 2963657339, flushing to timestamp: 2963658009.
Dropping flushed packet in player_put_packet, seqno 59586, timestamp 2963657691, flushing to timestamp: 2963658009.
requesting resend on 11 packets starting at 59576.
syncing to seqno 59588.
[  800.809592] WM8524: Sampling Rate=44100Hz
[  800.813679] WM8524: MCLK=16.934MHz
shairport
RTP: Packets out of sequence: expected: 59650, got 59651.
RTP: Packets out of sequence: expected: 59701, got 59724.
    CSeq: 6.
    DACP-ID: 78D9082BCC177063.
    Active-Remote: 1980701604.
    User-Agent: AirPlay/267.3.
    CSeq: 6.
    Audio-Jack-Status: connected; type=analog.
    Public: ANNOUNCE, SETUP, RECORD, PAUSE, FLUSH, TEARDOWN, OPTIONS, GET_PARAMETER, SET_PARAMETER.
RTP: Packets out of sequence: expected: 59845, got 59846.
    CSeq: 7.
    DACP-ID: 78D9082BCC177063.
    Active-Remote: 1980701604.
    User-Agent: AirPlay/267.3.
    CSeq: 7.
    Audio-Jack-Status: connected; type=analog.
    Public: ANNOUNCE, SETUP, RECORD, PAUSE, FLUSH, TEARDOWN, OPTIONS, GET_PARAMETER, SET_PARAMETER.
    RTP-Info: seq=60145;rtptime=2963859105.
    CSeq: 8.
    DACP-ID: 78D9082BCC177063.
    Active-Remote: 1980701604.
    User-Agent: AirPlay/267.3.
    CSeq: 8.
    Audio-Jack-Status: connected; type=analog.
Dropping flushed packet in player_put_packet, seqno 60145, timestamp 2963853593, flushing to timestamp: 2963859105.
Dropping flushed packet in player_put_packet, seqno 60146, timestamp 2963853945, flushing to timestamp: 2963859105.
    RTP-Info: seq=60147;rtptime=2963861753.
    CSeq: 9.
    DACP-ID: 78D9082BCC177063.
    Active-Remote: 1980701604.
    User-Agent: AirPlay/267.3.
Dropping flushed packet in player_put_packet, seqno 60147, timestamp 2963856241, flushing to timestamp: 2963861753.
Dropping flushed packet in player_put_packet, seqno 60148, timestamp 2963856593, flushing to timestamp: 2963861753.
    CSeq: 9.
    Audio-Jack-Status: connected; type=analog.
Dropping flushed packet in player_put_packet, seqno 60149, timestamp 2963856945, flushing to timestamp: 2963861753.
Dropping flushed packet in player_put_packet, seqno 60150, timestamp 2963857297, flushing to timestamp: 2963861753.
Dropping flushed packet in player_put_packet, seqno 60151, timestamp 2963857649, flushing to timestamp: 2963861753.
Dropping flushed packet in player_put_packet, seqno 60152, timestamp 2963858001, flushing to timestamp: 2963861753.
Dropping flushed packet in player_put_packet, seqno 60153, timestamp 2963858353, flushing to timestamp: 2963861753.
Dropping flushed packet in player_put_packet, seqno 60154, timestamp 2963858705, flushing to timestamp: 2963861753.
Dropping flushed packet in player_put_packet, seqno 60155, timestamp 2963859057, flushing to timestamp: 2963861753.
Dropping flushed packet in player_put_packet, seqno 60156, timestamp 2963859409, flushing to timestamp: 2963861753.
Dropping flushed packet in player_put_packet, seqno 60157, timestamp 2963859761, flushing to timestamp: 2963861753.
Dropping flushed packet in player_put_packet, seqno 60158, timestamp 2963860113, flushing to timestamp: 2963861753.
Dropping flushed packet in player_put_packet, seqno 60159, timestamp 2963860465, flushing to timestamp: 2963861753.
Dropping flushed packet in player_put_packet, seqno 60160, timestamp 2963860817, flushing to timestamp: 2963861753.
Dropping flushed packet in player_put_packet, seqno 60161, timestamp 2963861169, flushing to timestamp: 2963861753.
Dropping flushed packet in player_put_packet, seqno 60162, timestamp 2963861521, flushing to timestamp: 2963861753.
syncing to seqno 60163.
    CSeq: 10.
    DACP-ID: 78D9082BCC177063.
    Active-Remote: 1980701604.
    User-Agent: AirPlay/267.3.
    CSeq: 10.
    Audio-Jack-Status: connected; type=analog.
    Connection: close.
RTSP shutdown requested.
closin[  804.641867] WM8524: Sampling Rate=44100Hz
g RTSP connection.
shutting down RTP thread
[  804.658152] WM8524: MCLK=16.934MHz
rtp_timing_sender thread interrupted. terminating.
culling threads.
new RTSP connection.
RTSP shutdown requested.
closing RTSP connection.
terminating RTSP thread.
Audio receiver -- Server RTP thread interrupted. terminating.
Control RTP thread interrupted. terminating.
Timing RTP thread interrupted. terminating.
Closed and terminated timer requester thread.
Timing RTP thread terminated.
terminating RTSP thread.
```

Tested with iOS 9.2 and shairport-sync 2.6.

**Comment by joerg-krause** [comment]
Now I downloaded all files from the cloud and the issue happens again:

```
syncing to seqno 15374.
[ 2284.600942] WM8524: Sampling Rate=44100Hz
[ 2284.605026] WM8524: MCLK=16.934MHz
RTP: Packets out of sequence: expected: 15379, got 15380.
    CSeq: 37.
    DACP-ID: 3A17CF70D414D3F5.
    Active-Remote: 117596997.
    User-Agent: AirPlay/267.3.
    CSeq: 37.
    Audio-Jack-Status: connected; type=analog.
    Public: ANNOUNCE, SETUP, RECORD, PAUSE, FLUSH, TEARDOWN, OPTIONS, GET_PARAMETER, SET_PARAMETER.
    CSeq: 38.
    DACP-ID: 3A17CF70D414D3F5.
    Active-Remote: 117596997.
    User-Agent: AirPlay/267.3.
    CSeq: 38.
    Audio-Jack-Status: connected; type=analog.
    Public: ANNOUNCE, SETUP, RECORD, PAUSE, FLUSH, TEARDOWN, OPTIONS, GET_PARAMETER, SET_PARAMETER.
    CSeq: 39.
    DACP-ID: 3A17CF70D414D3F5.
    Active-Remote: 117596997.
    User-Agent: AirPlay/267.3.
    CSeq: 39.
    Audio-Jack-Status: connected; type=analog.
    Public: ANNOUNCE, SETUP, RECORD, PAUSE, FLUSH, TEARDOWN, OPTIONS, GET_PARAMETER, SET_PARAMETER.
    CSeq: 40.
    DACP-ID: 3A17CF70D414D3F5.
    Active-Remote: 117596997.
    User-Agent: AirPlay/267.3.
    CSeq: 40.
    Audio-Jack-Status: connected; type=analog.
    Public: ANNOUNCE, SETUP, RECORD, PAUSE, FLUSH, TEARDOWN, OPTIONS, GET_PARAMETER, SET_PARAMETER.
    CSeq: 41.
    DACP-ID: 3A17CF70D414D3F5.
    Active-Remote: 117596997.
    User-Agent: AirPlay/267.3.
    CSeq: 41.
    Audio-Jack-Status: connected; type=analog.
    Public: ANNOUNCE, SETUP, RECORD, PAUSE, FLUSH, TEARDOWN, OPTIONS, GET_PARAMETER, SET_PARAMETER.
+++++++++
    CSeq: 42.
    DACP-ID: 3A17CF70D414D3F5.
    Active-Remote: 117596997.
    User-Agent: AirPlay/267.3.
    CSeq: 42.
    Audio-Jack-Status: connected; type=analog.
    Connection: close.
RTSP shutdown requested.
closing RTSP connection.
shutting down RTP thread
Audio receiver -- Server RTP thread interrupted. terminating.
Control RTP thread interrupted. terminating.
Timing RTP thread interrupted. terminating.
rtp_timing_sender thread interrupted. terminating.
Closed and terminated timer requester thread.
Timing RTP thread terminated.
terminating RTSP thread.
```

**Comment by mikebrady** [comment]
Thanks. I'll see if I can reproduce it, but due to work commitments, it could be be about two weeks...

**Comment by joerg-krause** [comment]
This works:

```
RTSP shutdown requested.
closing RTSP connection.
shutting down RTP thread
Audio receiver -- Server RTP thread interrupted. terminating.
Control RTP thread interrupted. terminating.
Timing RTP thread interrupted. terminating.
rtp_timing_sender thread interrupted. terminating.
Closed and terminated timer requester thread.
Timing RTP thread terminated.
terminating RTSP thread.
culling threads.
one joined...
new RTSP connection.
```

This not:

```
RTSP shutdown requested.
closing RTSP connection.
shutting down RTP thread
Audio receiver -- Server RTP thread interrupted. terminating.
Control RTP thread interrupted. terminating.
Timing RTP thread interrupted. terminating.
rtp_timing_sender thread interrupted. terminating.
Closed and terminated timer requester thread.
Timing RTP thread terminated.
culling threads.
new RTSP connection.
RTSP shutdown requested.
closing RTSP connection.
terminating RTSP thread.
terminating RTSP thread
```

The problem in the second case is, that a new RTSP connection is requested before the RTSP thread is terminated.

**Comment by joerg-krause** [comment]
fwitw, selects in https://github.com/mikebrady/shairport-sync/blob/master/rtsp.c#L1630 returns with: `(-1): Interrupted system call`

**Comment by mikebrady** [comment]
That's great detective work, @joerg-krause, thanks.

**Comment by joerg-krause** [comment]
It looks to me like a race condition between the a RTSP shutting down, and creating a new RTSP thread. Unfortunatly, I'm not much experienced with pthreads.

**Comment by joerg-krause** [comment]
Okay, so much I understand so far. While one RTSP connection is shutting down, a new RTSP connection is accepted, a new thread created. The threads function `rtsp_conversation_thread_func` is calling `rtsp_read_request` and there the new thread is told to shut down: https://github.com/mikebrady/shairport-sync/blob/master/rtsp.c#L457. So, he does!

**Comment by joerg-krause** [comment]
The `EINTR` error code returned by `select()` (https://github.com/mikebrady/shairport-sync/issues/177#issuecomment-170871910) means that the system call was interrupt by a signal handler. Maybe something not to worry about?

**Comment by mikebrady** [comment]
I'm still investigating this – it's tricky.

**Comment by joerg-krause** [comment]
Indeed! Maybe the issue only appears on slow devices where a RTSP thread is shutting down and doing some cleanup while a new RTSP thread is created and starts running.

**Comment by mikebrady** [comment]
What's changed between 9.1 and 9.2, I think, is that in 9.2 the Music app tears down the play session and immediately starts another one each time you change track. The signalling Shairport Sync uses internally to shut down threads at the end of play session was, shutting down threads in the new play session too. The signalling method has been changed to be thread-specific. The problem is not completely fixed, but it's a lot better than it was before. If you're interested in trying it out, it's on the `better-stopping` branch.

**Comment by mikebrady** [comment]
The changes referred to above have been made part of stable release 2.8.0.

**Comment by moodeaudio** [comment]
Hi Mike,

Much better in 2.8 :-) 

(1) What I'm seeing is that iTunes and Apple Music on ElCap and IOS 9.2 seems to work perfectly except for Beats 1 (see #2 below). No problems switching tracks, playlists, etc. I set Moode to "Resume playback after Airplay" and this happens within 2-3 seconds after pausing iTunes play. It happens instantly after deselecting Airplay receiver from iTunes volume menu. All good.

(2) However, when using System volume Airplay on ElCap for example when streaming Youtube or some other non-iTunes source, or when playing iTunes Beats 1 station which requires use of System volume Airplay, the Airplay session often does not seem to be cleaned up when pause or switch playlist item. The message "Airplay speaker in use" pops up. When it does work cleanly it takes about 20 seconds for Moode to resume playback after pausing the source. More or less same behavior on IOS with non-iTunes sources.

I've also experienced some odd hang-like symptoms in the player (iTunes, Browser Youtube...) when using ElCap System volume Airplay and then subsequently option-click to open System volume control again.

It appears that in new OSX/IOS there is major difference in how Apple is handling Airplay within iTunes and when it is used outside of iTunes via System volume..

Regards,
Tim

**Comment by mikebrady** [comment]
Thanks Tim. I've continued to experiment with the stop/start transitions caused now by both iOS 9 and YouTube when played through AirPlay. I'm not sure I've actually nailed the problem, but I think I have improved it. Would you be willing to try the development version, just to see whether it has improved matters?

**Comment by moodeaudio** [comment]
Sure no prob :-) I'll run tests today. 
Do u have a binary?

**Comment by mikebrady** [comment]
Just sent one!

**Comment by moodeaudio** [comment]
Got it. I'll test in a bit.

**Comment by moodeaudio** [comment]
I'm not getting any audio from 2.9 dev. The "before" script is executed that stops MPD playback in Moode so at least its receiving signals from iTunes. Also "Airplay device in use" message still occurring.

**Comment by mikebrady** [comment]
Yikes, that's a bit unexpected. Let me do some checking.

**Comment by mikebrady** [comment]
Hi Tim. I just did a full recompile of the latest, just in case there was something fishy about the one I sent – I've just emailed it to you. I put it into a just-downloaded image of "Release: 2.5, 2016-02-DD". I was routing audio to the built-in audio and had the volume control set to both "software" and "hardware". Long story short – after a few restarts, it works for iTunes and, when used as a System Output, it works with Quicktime and with YouTube in Safari and Chrome. I don't subscribe to Apple Music, so I was only able to try Beats 1 and it works similarly. (BTW, I notice that the AirPlay icon disappears from iTunes when you are playing Beats, and it doesn't come back when you go back to "My Music" – that looks like a bug in iTunes.)
Trying it on iOS 9.2 on ain iPhone 6, it works with TuneIn Radio and with Music and the Radio Svizzera app. A mystery.

**Comment by mikebrady** [comment]
Hi Tim. I just did a full recompile of the latest, just in case there was something fishy about the one I sent, and I attache it here. I put it into a just-downloaded image of "Release: 2.5, 2016-02-DD". I was routing audio to the built-in audio and had the volume control set to both "software" and "hardware". Long story short – after a few restarts, it works for iTunes and, when used as a System Output, it works with Quicktime and with YouTube in Safari and Chrome. I don't subscribe to Apple Music, so I was only able to try Beats 1 and it works similarly. (BTW, I notice that the AirPlay icon disappears from iTunes when you are playing Beats, and it doesn't come back when you go back to "My Music" – that looks like a bug in iTunes.)
Trying it on iOS 9.2 on ain iPhone 6, it works with TuneIn Radio and with Music and the Radio Svizzera app.

**Comment by moodeaudio** [comment]
Hi Mike - Apple disables Airplay multiple receivers for Beats 1, probably because of some silly licensing thing.

I didn't see an attachment in any of the emails. Did u get a bounce back after sending?

**Comment by mikebrady** [comment]
Thanks for they for on the Beats 1 thing. No bounce-back. I'll send it again from a different email address.

**Comment by moodeaudio** [comment]
Hi Mike - 2.9 looking solid after stress testing. No issues whatsoever with connection handling :-)
-Tim

**Comment by mikebrady** [comment]
Thanks for the report, Tim. I'll be rolling these fixes into the next release.

**Comment by mikebrady** [comment]
This has been fixed in 2.8.1.
